### PR TITLE
Clean up after #7964

### PR DESCRIPTION
**Pull request by robertbastian**
Mainly addresses tests
* Splits tests between source data provider (locale-specific tests) and the component crate (behaviour-specific tests)
* Returns `Result` instead of panicking in constructor
* Tests edge cases like empty periods, single periods

#7969

## Changelog

N/A

### CODE REVIEW

File: provider/source/src/datetime/day_periods.rs [hunk 1]
```diff
@@ -174,18 +175,104 @@ mod tests {
...
+                .into_iter()
+                .collect()
+            )
+            .unwrap()
+            .unwrap()
+        )
+    }
+
+    #[test]
+    fn test_data() {
```
Comment (robertbastian) [inline provider/source/src/datetime/day_periods.rs]: it's unclear to me what the value of these tests is

**Comment by Manishearth** [comment]
won't be able to do a proper review for another week. I am fine with Shane reviewing this though, if I have any issues I'll bring them up post merge. Thanks for helping!

**Comment by robertbastian** [comment]
@sffc ?

File: components/datetime/src/provider/day_periods.rs [hunk 2]
```diff
@@ -199,8 +197,14 @@ impl DayPeriodRules {
...
 
-        let hour_periods =
-            hour_periods.map(|p| p.expect("Gap detected in 24-hour day period coverage"));
+        for p in &hour_periods {
+            if p.is_none() {
+                return Err("gap between periods");
+            }
+        }
+
+        #[allow(clippy::unwrap_used)] // just checked
```
Comment (sffc) [inline components/datetime/src/provider/day_periods.rs:208]: Nit: link to https://github.com/rust-lang/rust/issues/79711

```suggestion
        // TODO: change this to use array_try_map
        // <https://github.com/rust-lang/rust/issues/79711>
        #[allow(clippy::unwrap_used)] // just checked
```

File: components/datetime/src/provider/day_periods.rs [hunk 3]
```diff
@@ -242,122 +246,86 @@ impl DayPeriodRules {
...
+            ]
+            .into_iter()
+            .collect(),
+        )
+        .is_err());
     }
 
-    #[cfg(feature = "compiled_data")]
     #[test]
-    fn test_compiled_data() {
```
Comment (sffc) [inline components/datetime/src/provider/day_periods.rs:303]: Issue (optional if you offer justification): `test_compiled_data` is missing from the new code; you are testing the builder but not the end-to-end compiled data

Comment (robertbastian) [reply components/datetime/src/provider/day_periods.rs:303]: As explained in the PR description, I've split this into tests that test behaviour, here, and tests that test datagen, in datagen. I've then removed the datagen tests because it was just asserting that e.g. `zh` has a specific CLDR value

**Review by sffc — APPROVED** [review]

**Review by Manishearth — COMMENTED** [review]
thanks
